feat: add device group page

patriciareinoso opened the pull request:
# Description

This PR add a Device Group page to the NMS according to https://docs.google.com/document/d/11uhHKXsIVrzSTOrxFsSb0rFvNN4cNnvlxYH7DYGtUQY/edit?usp=sharing

## Navigation 
![image](https://github.com/user-attachments/assets/add48530-01ad-454e-b3dc-ef79d1d90327)

The suggested one did not look good because the size of the icon is too small and we couldnt see the details:
![image](https://github.com/user-attachments/assets/b0b0d170-2b85-46c5-b4e7-fc0a3cc79e60)

## New table

![image](https://github.com/user-attachments/assets/3f6301ec-eef5-4c92-9016-d964f7bd9347)

## Create DG modal
- Include new QoS fields: 5QI and ARP. Default values 1 and 6.
- Add dropdown to select NS

![image](https://github.com/user-attachments/assets/e9385905-fd5a-49f9-861e-a523a013cc46)

## Edit DG modal
- Existing values are prefilled
- Include new QoS fields: 5QI and ARP
- Name and network slice cannot be edited

![image](https://github.com/user-attachments/assets/35c504a5-2e06-4c2d-a521-a897d22df12d)

## Empty state no network slice 

![image](https://github.com/user-attachments/assets/266f152d-0bd8-47f7-a449-663b97f9f162)


## Empty state no device groups

![image](https://github.com/user-attachments/assets/46b6bed1-660b-41af-9fd1-59814a03c425)


## Compute Qos Values
- 5QI allowed values : 1, 2, 9
- ARP  range [ 1, 15 ]

According to [23.501 3GPP specification](https://www.etsi.org/deliver/etsi_ts/123500_123599/123501/16.06.00_60/ts_123501v160600p.pdf) Table 5.7.4-1 

5QI == 1  ==> PDB = 100 && PELR = 2 
5QI == 2  ==> PDB = 150 && PELR = 3 
5QI == 9  ==> PDB = 300 && PELR = 6 

## Code changes
- Add empty state:
   - No network slices available:  button to redirect to NS page.
   - No device groups available:  button to create a new DG.
- Create a new Device Group page which holds the Device Group table.
- Reorganize `components/DeviceGroupModal.tsx` into `app/(nms)/device-groups/modals.tsx`
-  `app/(nms)/device-groups/modals.tsx` contains:
    - Common modal `DeviceGroupModal`, which is "inherited" by `CreateDeviceGroupModal` and `EditDeviceGroupModal`
    - `DeleteDeviceGroupButton` which show the delete modal depending on whether the DG contains subscribers.
- Add a `Device Groups` entry in the navigation menu
- Remove code related to DGs from the network slice pages
- Modify `components/PageContent.tsx` and `components/PageHeader.tsx` to make their size configurable. Given the number of fields in the DG table, we need a wider table.
- Create a DeviceGroup type
- Group `utils/createDeviceGroup.tsx`, `utils/deleteDeviceGroup.tsx`, `utils/editDeviceGroup.tsx`, `utils/getDeviceGroup.tsx` into `utils/deviceGroupOperations.tsx` 
-  Create 2 types of error: `WebconsoleApiError` and `OperationError`. This is useful to show appropriate error message to the user.
- Remove the automatic creation of a DG when a NS is created.

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added tests that validate the behaviour of the software.
- [ ] I validated that new and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

gruyaume commented on app/(nms)/device-groups/page.tsx:
```diff
@@ -0,0 +1,129 @@
+            { content: "Subscriber IP Pool" },
+            { content: "DNS" },
+            { content: "MTU" },
+            { content: "MBR Downstream" },
+            { content: "MBR Upstream" },
```
I would avoid these unnecessary acronyms in favor of 
```
Bitrate Downstream
Bitrate Upstream
```

gruyaume replied:
We should specify the unit (Mbps)

gruyaume commented on app/(nms)/device-groups/modals.tsx:
```diff
@@ -0,0 +1,521 @@
+  networkSlice: Yup.string()
+      .min(1)
+      .required("Network slice is required."),
+  ueIpPool: Yup.string()
+      .required("IP is required")
+      .matches(regexpCIDR, "Invalid IP Address Pool."),
```
```suggestion
  ueIpPool: Yup.string()
      .required("IP pool is required")
      .matches(regexpCIDR, "Invalid IP Address Pool."),
```

gruyaume commented on app/(nms)/device-groups/modals.tsx:
```diff
@@ -0,0 +1,521 @@
+const regexpCIDR =
+  /^((25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\/([1-9]|[1-2][0-9]|3[0-2])$/;
+
+const validQCIValues = [1, 2, 9];
```
Are we using qci because we can't have an int in the middle of the variable? `5qi` would be the appropriate concept in 5g. We can also use `fiveQi` or something along those lines if that's the issue

gruyaume commented on app/(nms)/device-groups/modals.tsx:
```diff
@@ -0,0 +1,521 @@
+  ueIpPool: string;
+  dns: string;
+  mtu: number;
+  MBRDownstreamMbps: number | null;
+  MBRUpstreamMbps: number | null;
```
Why accept null values for those two?

patriciareinoso replied:
There is no default value for these 2 (only a hint/placeholder in the Create modal). We need to pass an initial value to the modal form, that's why we accept null.
I took this from the existing modal before my modifications.

gruyaume commented on app/(nms)/device-groups/page.tsx:
```diff
@@ -0,0 +1,129 @@
+            { content: "5QI" },
+            { content: "ARP" },
+            { content: "Actions"},
+            { content: ""},
```
Why this last column?

patriciareinoso replied:
The buttons to edit and delete keep showing up one below the other. The solution I found was to put each one in a different cell, which makes one of the header cells empty

gruyaume commented on app/(nms)/device-groups/page.tsx:
```diff
@@ -0,0 +1,129 @@
+            { content: "Network Slice" },
+            { content: "Subscriber IP Pool" },
+            { content: "DNS" },
+            { content: "MTU" },
```
We should specify the unit

patriciareinoso replied:
I tried several combinations trying to display the units : i decided for this one 
![image](https://github.com/user-attachments/assets/eea453dd-dd19-4c0f-aeda-991822bae380)
WDYT?

gruyaume commented on components/NetworkSliceTable.tsx:
(not really related to this file)
We should probably stop creating a device group automatically when we create a network slice. I'm happy for this change to be in a different PR.

patriciareinoso replied:
Done

gruyaume reviewed (changes requested):
Overall this looks good. My comments are around making concepts clearer to users.

patriciareinoso commented:
@gruyaume as discussed:
- I added 2 empty states, one for the no NS case and one for the no DG case.
- Set the default value of QCI to 1 in the create DG modal
- Added the units in table headers

gruyaume commented on app/(nms)/device-groups/page.tsx:
```diff
@@ -0,0 +1,237 @@
+        <br></br>
+      </PageHeader>
+      <PageContent colSize={8}>
+        <EmptyState image={""} title="No device groups available">
```
```suggestion
        <EmptyState image={""} title="No device group available">
```

gruyaume commented on app/(nms)/device-groups/page.tsx:
```diff
@@ -0,0 +1,237 @@
+            To create a device group, first create a network slice.
+          </p>
+          <Button appearance="positive" onClick={() => router.push("/network-configuration")}>
+            Go to network slices page
```
```suggestion
            Go to the "Network Slices" page
```

gatici commented on app/(nms)/device-groups/modals.tsx:
```diff
@@ -0,0 +1,521 @@
+}
+
+const regexIp =
+  /^(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)$/;
+
+const regexpCIDR =
+  /^((25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\/([1-9]|[1-2][0-9]|3[0-2])$/;
```
This regex allow subnet masks from 1-32, however 0 is also a valid subnet mask. It also does not include the logic that IP range is valid for given subnet mask.

gatici replied:
This regex accepts the leading zeros in the IP sections. For example 001.020.030.030/24 is acceptable according to this regex. 
To be honest, these regexes are quite complex. Can we benefit from some JSX libraries ?

gatici reviewed (changes requested):
I tested the PR and it works well except one issue. When we create our first device group, it does not appear unless we refresh the page manually. This is breaking an existing functionality.

patriciareinoso commented:
> I tested the PR and it works well except one issue. When we create our first device group, it does not appear unless we refresh the page manually. This is breaking an existing functionality.

I tried to reproduce this several times and I never saw this behavior. I added a delay before invalidating the queries because I saw a race condition in the DELETE operation, so maybe something similar is happening for you. Can you try again ?
